azure-iot-common/azure/iot/common/asyncio_compat.py

```diff
@@ -0,0 +1,111 @@
+import asyncio
+import concurrent.futures
+
+
+def get_running_loop():
+    """
+    Helper function to use the best method to get an event loop to run async code
+    in.  This is a "best effort" function.  get_running_loop is preferred because it
+    works in more cases (especially if customers are using custom event loop policies),
+    but get_event_loop works well in simple cases (e.g. there is only one loop).
+    """
+    try:
+        loop = asyncio.get_running_loop()
+    except AttributeError:
+        loop = asyncio.get_event_loop()
+    return loop
+
+
+def create_task(coro):
+    """
+    "Best effort" function. If avaialable (Python 3.7+), use asyncio.create_task,
+    which is preferred as it is more specific for the goal of immediately scheduling
+    a task from a coroutine. If not avaialble, use the more general puprose 
+    asyncio.ensure_future
+    """
+    try:
+        task = asyncio.create_task(coro)
+    except AttributeError:
+        task = asyncio.ensure_future(coro)
+    return task
+
+
+def run(coro, debug=False):
+    """
+    Helper function to abstract event loop setup/cleanup. Uses asyncio.run() if it is
+    available. Otherwise, manually gets, runs and closes the event loop
+    """
+    try:
+        asyncio.run(coro, debug=debug)
+    except AttributeError:
+        loop = asyncio.get_event_loop()  # NOT running loop
+        loop.set_debug(debug)
+        loop.run_until_complete(coro)
+        loop.close()
+
+
+def create_future(loop):
+    """
+    Helper function to create a Future object.  Uses loop.create_future() if it is
+    available.  Otherwise, create the object directly.  loop.create_future is preferred
+    because it allows third parties to provide their own Future object, but it is only
+    available in 3.5.2+
+    """
+    try:
+        future = loop.create_future()
+    except AttributeError:
+        future = asyncio.Future(loop=loop)
+    return future
+
+
+def emulate_async(fn):
+    """
+    Apply as a decorator to emulate async behavior with a sync function/method
+    via usage of multithreading
+    """
+
+    async def async_fn_wrapper(*args, **kwargs):
+        loop = get_running_loop()
+
+        # Run fn in default ThreadPoolExecutor (CPU * 5 threads)
```

Comment:
>Run fn in default ThreadPoolExecutor (CPU * 5 threads) [](start = 10, length = 54)

better to mention this somewhere as doc string.

Reply:
No. Implementation details are not recommended for inclusion in docstrings

---
In reply to: [246960811](https://github.com/Azure/azure-iot-sdk-python-preview/pull/39#discussion_r246960811) [](ancestors = 246960811)

Reply:
my mistake...i did not intend to mention the CPU * 5 threads part as doc string...but was referring to the fact that we are using `None` as the executor parameter for `run_in_executor`.  That can be explained as we are using default executor and not a processpool or threadpool..

---
In reply to: [246964302](https://github.com/Azure/azure-iot-sdk-python-preview/pull/39#discussion_r246964302) [](ancestors = 246964302,246960811)

Reply:
Since our exact implementation of emulate_async may change (see: limitations of our current model), I don't consider us tied to using the loop's default executor. 

---
In reply to: [246979670](https://github.com/Azure/azure-iot-sdk-python-preview/pull/39#discussion_r246979670) [](ancestors = 246979670,246964302,246960811)